### ISSUE TITLE
Periodic issue with TotalConnect

### ISSUE THREAD

**Issue by FedgarLurch** [craigjmidwinter/total-connect-client#77]
Every... 5th time or so I do a restart on HA, I run into this error

```2020-04-16 08:56:38 ERROR (MainThread) [homeassistant.setup] Error during setup of component totalconnect
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 174, in _async_setup_component
    component.setup, hass, processed_config  # type: ignore
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/totalconnect/__init__.py", line 37, in setup
    client = TotalConnectClient.TotalConnectClient(username, password)
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 75, in __init__
    self.authenticate()
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 139, in authenticate
    self.populate_details(response)
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 187, in populate_details
    self.get_zone_details(location_id)
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 415, in get_zone_details
    + ', {"int": ["1"]}, 0)'
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 97, in request
    return self.request(request, attempts)
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 97, in request
    return self.request(request, attempts)
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 97, in request
    return self.request(request, attempts)
  [Previous line repeated 7 more times]
  File "/usr/local/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 120, in request
    "total-connect-client could not execute request.  Maximum attempts tried."
Exception: total-connect-client could not execute request.  Maximum attempts tried.
```

Ill have to restart it 3 to 5 times before it goes away. 

I have noticed when it does work, it takes almost 70 seconds to setup. I have a large facility with multiple panels, and many sensors (not sure if that matters)

Panel 1:
16 sensors
Panel 2: 
23 sensors

Let me know if I can help by adding more detail. 
EDIT: (System info)
HassOS 3.12 Version 217
HA Version 0.107.7

Thanks!

**Comment by austinmroczek** [comment]
Which version of HA are you running?

**Comment by FedgarLurch** [comment]
Sorry. 0.107.0 I noticed it happening back at 0.103.0 as well though. Ive added my system info above^

**Comment by austinmroczek** [comment]
I will try to look at it later tonight or this weekend. Could be an issue on the Honeywell end, or internet between you and them.  

The way it’s written right now, any time a request is tried, if we can’t connect it tries again in 3 seconds.  After 10 tries it gives the error message you have.  

3 and 10 are arbitrary numbers I put in without too much thought.  Maybe changing those will help.

I think this was put in at 0.103 so that’s why you started seeing them. Likely before updates were silently failing.  Perhaps we need to go back to that instead of throwing the error.

**Comment by austinmroczek** [comment]
Are you only getting this on startup?

**Comment by austinmroczek** [comment]
I don't know if this is related or not, but I also had hundreds of "Maximum attempts tried" entries in my log 3 days ago.  Maybe the TotalConnect service had a hiccup?  @FedgarLurch have you continued to have issues in the last few days?

My system only has one panel but is taking roughly 5-10 seconds to get started up.  I'm working to add some time tracking features into the code.  Hopefully good enough to post later today/tonight.

**Comment by austinmroczek** [comment]
@FedgarLurch sorry I was really slow getting back to you.  I've added some timers inside the code that will track which functions are slow.  

Please do the following your raspbian command prompt:
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/TotalConnectClient.py`
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/info.py`
`python3 info.py username password -1`

Please remove any sensitive data, especially username and password which are near the top of the output, and then post here.

**Comment by FedgarLurch** [comment]
@austinmroczek Sorry about the delay, I meant to respond. 
I only have issues with on restart. I did have an issue 2 days ago, but generally its been fine. 
That being said I don't know if its been having issues because I have been on the road. I can take a look at the logs as soon as I am home.

**Comment by austinmroczek** [comment]
The new code I posted measures the speed of the various actions we do at startup and should give us an idea if it is the size of your system that is affecting things.  Testing my system I'm getting about 5 seconds average startup time, with a few outliers over 10 seconds.

If we can confirm it's just a startup time issue, I think there is a way to tell Home Assistant to wait longer during the startup process.  Otherwise we'll need to keep troubleshooting.

**Comment by dulitz** [comment]
I am experiencing this problem right now but was not last night. I see the problem when constructing a new TotalConnectClient object.

Their WSDL endpoint is up (I can fetch it with wget). LoginAndGetSessionDetails succeeds. But then for GetZonesListInStateEx_V1 I get a response.ResultCode of self.CONNECTION_ERROR for all 10 attempts.

This happens both with the latest version on PyPI and with the current code on GitHub.

**Comment by austinmroczek** [comment]
@dulitz 
I just test my own system and am not having trouble.  

Please follow the troubleshooting instructions (on the README / main page) and post your info here.  

Are you getting any other signs of network trouble?  My network and/or Pi that's running the code sometimes clunks out...then later fixes itself, but I usually see other services having trouble too (as seen in my Home Assistant logs) so I know it's a network-wide problem.

**Comment by austinmroczek** [comment]
I'll try to spend some time on this over the weekend.

A possible work-around would be to create two TotalConnect users, with access to one location each.  This might also help confirm that it's one panel/location causing the problems.

Can you tell exactly which step is giving you problems?

Our current flow is
1. LoginAndGetSessionsDetails
2. GetZonesListInStateEx_V1
3. GetPanelMetaDataAndFullStatus  
4. Wait then repeat step 3 (forever and ever)

I can look at splitting LoginAndGetSessionDetails into calls for AuthenticateUserLogin and then GetSessionDetails.  However, I think this is slightly less efficient traffic-wise.  Also, I don't know for sure, but I think step 1 is just a database lookup on the server.

**Comment by FedgarLurch** [comment]
@austinmroczek Hey! Me again, Its actually been working pretty well of late, but this morning I was doing some updating (Everything is completely up to date as of this morning) and while restarting, I got the same error as above (Pasted below just incase the details are necessary.)

Let me know how I can help you troubleshoot. I have discord if that is faster, or just leave me some instructions and I will do what I can. I know my install is much larger than most since I have multiple buildings with multiple panels. 

`Logger: homeassistant.config_entries
Source: components/totalconnect/__init__.py:29
First occurred: 11:18:34 AM (1 occurrences)
Last logged: 11:18:34 AM

Error setting up entry Total Connect for totalconnect
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 293, in async_setup
    result = await component.async_setup_entry(hass, self)  # type: ignore
  File "/usr/src/homeassistant/homeassistant/components/totalconnect/__init__.py", line 29, in async_setup_entry
    client = await hass.async_add_executor_job(
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 95, in __init__
    self.authenticate()
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 205, in authenticate
    self.populate_details(response)
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 292, in populate_details
    self.get_zone_details(location_id)
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 496, in get_zone_details
    result = self.request(
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 157, in request
    return self.request(request, attempts)
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 157, in request
    return self.request(request, attempts)
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 157, in request
    return self.request(request, attempts)
  [Previous line repeated 7 more times]
  File "/usr/local/lib/python3.8/site-packages/total_connect_client/TotalConnectClient.py", line 179, in request
    raise Exception(
Exception: total-connect-client could not execute request.  Maximum attempts tried.`

**Comment by dulitz** [comment]
@FedgarLurch you said you have multiple panels in multiple structures. Do you get this error on some panels more than others?

I have also seen that, when this error is occurring, making requests too quickly can cause it to fail to authenticate. Hypothesis: The reason we fail in populate_details() is due to some transient panel/network/server issue, while the reason we fail authentication is due to rate limits (or a transient server issue, but often rate limits).

That is yet another reason why I think authentication and populate_details() should be in two different calls -- authentication is expensive and, when populate_details() is failing, retrying the whole thing doubles our number of requests.

What happens now:
`  success(auth), fail(populate), success(auth), fail(populate), ... fail(auth), ...`
where we eventually fail auth due to rate limits and then we never recover.

What should happen:
`   success(auth), fail(populate), fail(populate), ... success(populate)`
where populate eventually succeeds as soon as the transient panel/network/server issue resolves.

**Comment by dulitz** [comment]
@austinmroczek 

> I can look at splitting LoginAndGetSessionDetails into calls for AuthenticateUserLogin and then GetSessionDetails.

After looking at the stack trace from @FedgarLurch and my own, I don't think that would help. I think I know what would help and will send a PR after you land get_partitions.

**Comment by austinmroczek** [comment]
@dulitz 

Not sure your plan, but on a discord with FedgarLurch I discussed this:

Right now the code tries a request, and if it fails, waits three seconds then tries again, for up to 10 tries.  Three and ten were arbitrary numbers I threw in there.  

If there is a rate-limit of some type on the server, we might want to try increasing the delay on each retry.  I was thinking of just adding three more to the delay each time, or maybe doubling the delay each time.  

I'm hoping your plan is better.

**Comment by dulitz** [comment]
My idea is orthogonal, which is good because we can do them both and have 2
shots on goal. :)

**Comment by austinmroczek** [comment]
I think we're going to call this fixed with #123 thanks to @dulitz 

If you want to try the new code immediately, download it from here at Github.  I'll make a new pypi package soon.

If folks continue to have issues, please post a new issue.

**Comment by dulitz** [comment]
Just for completeness: with the new code, if/when this happens again, it will happen when the TotalConnectClient.locations property is accessed.

Please send the full backtrace from the exception. Also, if you can, please sign in to the TotalConnect app or website and attempt to sync the panel.

What I _expect_ to happen is that get_partition_details() will be the call that usually fails, and that when it fails you will also not be able to sync the panel on the TotalConnect app or website, and both those failures will be because the panel cannot connect to the TotalConnect servers (via Internet or radio or however it is usually configured).

We will see how closely reality conforms to my expectations. :)
